### ISSUE TITLE
Mixing virtual and physical registers may produce bugs

### ISSUE THREAD

**Issue by Maratyszcza** [Maratyszcza/PeachPy#30]
If virtual register is allocated before physical registers, PeachPy can bind it to the same physical register.

**Comment by jhogg41** [comment]
Getting bitten by this one particularly annoying re DIV and MUL where you need to use specific physical registers, which GeneralPurposeRegister() has usefully assigned to other things.

Would be useful if I could specify an except clause to the Function(), and then have _bind_registers() avoid allocating it to something else.

**Comment by Maratyszcza** [comment]
@jhogg41 Do you have a repro? It would help a lot

**Comment by jhogg41** [comment]
Please find attached - if you delete the myrax, myrcx and myrdx on lines
120-123 of gemm_py.py and search replace them for the physical register
names, you'll see that some of rax, rcx and rdx get used for function
arguments, and then when we get down to the DIV before above the "Main
loop" declaration we need physical rax and rdx.

I've also attached a code that calls the output, but as the code is
somewhat experimental it may well not get the right answer even when
register assignment issues are resolved.

[NB: code itself is mainly me playing about trying to get an automatic
instruction scheduler for the inner loop, as a first step towards an end
goal of trying to derive asm from the same DSL as a higher level DAG-based
runtime system code generator]

Thanks for the project!

Jonathan.

**Comment by Maratyszcza** [comment]
I don't see the attachment. Probably GitHub stripped it off.

**Comment by jhogg41** [comment]
Will send via email direct to you.

**Comment by xcombelle** [comment]
I was just wondering how one could be sure that in the following function:
```
x = Argument(int64_t)
y = Argument(int64_t)

with Function("DotProduct", (x, y), int64_t) as asm_function:
    reg_x = GeneralPurposeRegister64()
    LOAD.ARGUMENT(reg_x, x)
    LOAD.ARGUMENT(rax, y)
    MUL( reg_x)
    RETURN(rax)
```

which as the following assembly

```
	MOV rax, rsi
	MUL rdi
	RET
```

how one can be sure that rax was not assigned to reg_x.

Maybe a possible syntax would be

```
x = Argument(int64_t)
y = Argument(int64_t)
with Function("DotProduct", (x, y), int64_t) as asm_function:
    reg_x = GeneralPurposeRegister64()
    reg_rax = GeneralPurposeRegister64(reg=rax)    
    LOAD.ARGUMENT(reg_x, x)
    LOAD.ARGUMENT(reg_rax, y)
    MUL( reg_x)
    RETURN(reg_rax)
```

A related question made me wonder how to be sure that the assigned register don't mess with the function call abi for example overwriting the stack pointer or the program pointer or a parameter

**Comment by Maratyszcza** [comment]
@xcombelle Normally PeachPy would detect dependencies between instructions and make sure registers are allocated in non-conflicting way, but something got broken

**Comment by xcombelle** [comment]
@Maratyszcza thanks

**Comment by Maratyszcza** [comment]
I believe the problem is fixed 8f742b7911691ffa2f4ca95370677838856676a2. Please re-open if you find other buggy cases.
